src/spatial.rs

```diff
@@ -45,24 +45,35 @@ pub fn base(dims: &[usize]) -> bool {
     dims.iter().all(|&x| x == 2)
 }
 
-fn _next_shapes(dims: &[usize]) -> Vec<Vec<usize>> {
+fn _next_shapes_up(dims: &[usize]) -> Vec<Vec<usize>> {
     let mut results = Vec::new();
-
     if dims.iter().all(|&x| x == 2) {
         let mut up = dims.to_vec();
         up.push(2);
```

Comment:
Don't use mutation here. Use the same chain trick as elsewhere in the file. Make a helper of that and use it